addons/udes_stock/models/stock_inventory.py

```diff
@@ -24,19 +25,94 @@ def action_done(self):
         Extends the parent method by ensuring that there are no
         incomplete preceding inventories.
 
+        Also checks that a user is allowed to adjust reserved stock.
+
         Raises a ValidationError otherwise.
         """
+        User = self.env['res.users']
+
         for prec in self.u_preceding_inventory_ids:
             if prec.state != 'done':
                 raise ValidationError(
                     _('There are undone preceding inventories.'))
 
+        if self._is_adjusting_reserved():
+            warehouse = User.get_user_warehouse()
+            if not (warehouse.u_inventory_adjust_reserved or
+                    self.env.user.has_group("udes_security.group_debug_user")):
+                raise ValidationError(_('Adjusting reserved stock is not allowed.'))
+
         return super(StockInventory, self).action_done()
 
+    @api.multi
+    def button_done(self):
+        """Add a popup to inform a user that they are adjusting reserved stock."""
+        self.ensure_one()
+
+        if self._is_adjusting_reserved():
+            return {
+                'name': _('Adjust Reserved Stock?'),
+                'type': 'ir.actions.act_window',
+                'view_type': 'form',
+                'view_mode': 'form',
+                'res_model': 'stock.inventory',
+                'views': [(self.env.ref('udes_stock.view_adjust_reserved').id, 'form')],
+                'view_id': self.env.ref('udes_stock.view_adjust_reserved').id,
+                'target': 'new',
+                'res_id': self.id,
+                'context': self.env.context,
+            }
+        else:
+            return self.action_done()
+
+    def _is_adjusting_reserved(self):
+        """Check if a user is adjusting reserved stock."""
+        self.ensure_one()
+        for line in self.line_ids:
+            if line.reserved_qty and line.theoretical_qty != line.product_qty:
+                return True
+        return False
+
     @api.multi
     def write(self, values):
         if 'done' in self.mapped('state'):
             raise UserError(
                 _('Cannot write to an adjustment which has already been '
                   'validated'))
         return super(StockInventory, self).write(values)
+
+
+class StockInventoryLine(models.Model):
+    _name = 'stock.inventory.line'
+    _inherit = 'stock.inventory.line'
+
+    reserved_qty = fields.Float(
+        'Reserved Quantity',
+        compute='_compute_reserved_qty',
+        digits=dp.get_precision('Product Unit of Measure'),
+        readonly=True,
+        store=False
+    )
+
+    @api.one
+    @api.depends(
+        'location_id',
+        'product_id',
+        'package_id',
+        'product_uom_id',
+        'company_id',
+        'prod_lot_id',
+        'partner_id'
+    )
+    def _compute_reserved_qty(self):
+        """Compute the reserved quantity for the line."""
+        if not self.product_id:
+            self.reserved_qty = 0
```

Comment:
is there a unit test that covers this path?

Reply:
I don't think we will hit this bit of logic here as product_id is required so I'm going to remove it. I'll add tests for this function anyway though